File: lab05_transactions/src/simple/credits_and_debits_producer.py
#!/usr/bin/env python3
import sys
import os
sys.path.append(os.path.abspath('../common'))
from bank_transaction_generator import BankTransactionGenerator
from confluent_kafka import Producer

# todo Configure the producer
# What is required for transactions
props = {}
p = Producer(props)
CREDITS_TOPIC = "credits"
DEBITS_TOPIC = "debits"
try:
    # todo setup the producer for using transactions

    for transaction in BankTransactionGenerator(1):
        p.begin_transaction()
        suspicious = " <-Suspicious!" if transaction["suspicious"] else ""

        # todo produce two messages: one to the credits-topic and one to the debits topic
        # Mark the suspicious transactions by appending the variable above

        print("{} -> {}: {}€".format(transaction["sender_account"], transaction["receiver_account"], transaction["amount"]))
        # For this example we need to flush the messages manually, otherwise we might not see any aborted transactions in the topic
        p.flush()

        if transaction["suspicious"]:
            # todo abort the transaction when it is suspicious

            print("Suspicious transaction between {} and {}! Amount: {}€",
                  transaction["sender_account"], transaction["receiver_account"], transaction["amount"])
        else:
            # todo otherwise commit it
            pass
finally:
    p.flush()

File: lab05_transactions/src/main/java/example/simple/CreditsAndDebitsProducer.java
package example.simple;

import example.common.BankTransaction;
import example.common.BankTransactionSupplier;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.List;
import java.util.Properties;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class CreditsAndDebitsProducer {
    public static void main(final String[] args) {
        final Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        // We have to assign a transactional ID so that Kafka knows who is responsible for the transaction.
        // This must be one-to-one! One ID per producer!
        // props.put(??, ??);

        final String CREDITS_TOPIC = "bank-credits";
        final String DEBITS_TOPIC = "bank-debits";

        int numMessages = 50;
        if (args.length == 1) {
            numMessages = Integer.parseInt(args[0]);
        }


        // Here we get a list of transactions (.limits the length of the stream).
        // Some of them are treacherous
        final List<BankTransaction> toGreet = Stream.generate(new BankTransactionSupplier(10000)).limit(numMessages).collect(Collectors.toList());

        try (Producer<String, String> producer = new KafkaProducer<>(props)) {
            // Before we can use transactions, we need to initialize them
            // todo

            for (BankTransaction transaction : toGreet) {
                // Let's start the transaction
                // How?

                // We simply append "<- Suspicious!" to a transaction value to better see when such transactions come through.
                String suspicious = transaction.isSuspicious ? " <-Suspicious!" : "";
                ProducerRecord<String, String> senderRecord = new ProducerRecord<>(CREDITS_TOPIC, transaction.sender_account, transaction.amount + "€" + suspicious);
                ProducerRecord<String, String> receiverRecord = new ProducerRecord<>(DEBITS_TOPIC, transaction.receiver_account, transaction.amount + "€" + suspicious);

                // Let's send the records
                producer.send(senderRecord);
                producer.send(receiverRecord);
                producer.flush();

                // For illustrative purposes only:

                if (transaction.isSuspicious) {
                    System.out.println("Suspicious transaction between " + transaction.sender_account + " and " + transaction.receiver_account + "! Amount: " + transaction.amount);
                    // todo cancel transaction
                } else {
                    // Todo commit transactions
                }
            }
            // That's not quite true ;)
            System.out.println("Produced " + numMessages + " Messages");
        }
    }
}
